Maya: Implement Hardware Renderer 2.0 support for Render Products

BigRoy opened the pull request:
## Brief description

This initiates the implementation supporting Hardware Renderer 2.0 in OP 3.0 as per #2126 and uses logic from #2113 as a basis.

I had to make a tweak compared to the code in #2113 and be explicit in passing the camera along with the Render Product otherwise I got this error:
```python
# Error: expected string or buffer
# Traceback (most recent call last):
#   File "<maya console>", line 10, in <module>
#   File "S:\openpype\OpenPype\openpype\hosts\maya\api\lib_renderproducts.py", line 399, in get_files
#     force_cameras=[product.camera]
#   File "S:\openpype\OpenPype\openpype\hosts\maya\api\lib_renderproducts.py", line 354, in _generate_file_sequence
#     (R_SUBSTITUTE_CAMERA_TOKEN, self.sanitize_camera_name(cam)),
#   File "S:\openpype\OpenPype\openpype\hosts\maya\api\lib_renderproducts.py", line 230, in sanitize_camera_name
#     return re.sub('[^0-9a-zA-Z_]+', '_', camera)
#   File "C:\Program Files\Autodesk\Maya2020\bin\python27.zip\re.py", line 155, in sub
#     return _compile(pattern, flags).sub(repl, string, count)
# TypeError: expected string or buffer # 
```

## Testing notes:

Here's how I quickly tested whether it was able to retrieve the filenames:

```python
import openpype.hosts.maya.api.lib_renderproducts as rp

# Set renderer to hardware renderer 2.0
cmds.setAttr("defaultRenderGlobals.currentRenderer", "mayahardware2", type="string")

# Get the file list for the master layer
renderer = rp.get("defaultRenderLayer")
for product in renderer.get_render_products():
    print(renderer.get_files(product))
```

---

Aside of implementing this code what else are we expecting might be needed to get it operational in OP 3.0? For example the CreateRender currently will fail when the renderer is set to `mayahardware2` with:
```python
Traceback (most recent call last):
  File "S:\openpype\OpenPype\openpype\tools\creator\window.py", line 430, in _on_create
    data={"variant": variant}
  File "S:\openpype\OpenPype\repos\avalon-core\avalon\pipeline.py", line 924, in create
    instance = plugin.process()
  File "<string>", line 203, in process
  File "<string>", line 433, in _set_default_renderer_settings
KeyError: u'mayahardware2'
```

Maybe @tokejepsen can provide details as what he'd need for this to complete? Should it just allow render submissiosn to Deadline with publishing? Or are we expecting more?

tokejepsen commented:
Submission to Deadline would suffice for us.

mkolar commented:
@BigRoy Do you think this is testable?

BigRoy commented:
I think it might still need support in the **Create Render** logic and alike to not fail in that area and allow deadline submissions to happen. Other than that I think the basics are there. However, Create Render and the render logic is getting an overhaul now with #3097 - so it might be worth waiting on that first.

antirotor commented:
> I think it might still need support in the **Create Render** logic and alike to not fail in that area and allow deadline submissions to happen. Other than that I think the basics are there. However, Create Render and the render logic is getting an overhaul now with #3097 - so it might be worth waiting on that first.

I've created [PR to this branch](https://github.com/BigRoy/OpenPype/pull/6) with small changes that allows to use it on "current" system.

BigRoy commented on openpype/hosts/maya/plugins/publish/collect_render.py:
```diff
@@ -326,8 +326,8 @@ def process(self, context):
                                               layer=layer_name)),
-                "renderer": self.get_render_attribute("currentRenderer",
-                                                      layer=layer_name),
+                "renderer": self.get_render_attribute(
+                    "currentRenderer", layer=layer_name).lower(),
```
I'm a bit worried about forcing this to lowercase here - what if the render submission to a render manager requires the original Renderer name (including uppercase characters?) for the submission.

Deadline's MayaBatch plug-in seems to check against lowercase (except for a single `self.Renderer == "redshiftExport"` check which I think  is just a bug in Deadline's code since other mentions do use lowercase for it too.)

Would we possibly run ourselves into issues due to us forcing lowercase?

antirotor replied:
yeah, I was looking into it too. I think we are using renderer name in lower case everywhere and this particular case was the one and only that wasn't using lower case name of renderer. I vote for addressing this in another issue if necessary - this works on Deadline but we might hit problems when using other farms like RoyalRender or Muster where the renderer name can be completely different or case sensitive and it that case it is easier to create mapping between our lower case renderer and farm specific names.

BigRoy commented:
> I've created https://github.com/BigRoy/OpenPype/pull/6 with small changes that allows to use it on "current" system.

@antirotor Thanks - I've merged it. I have noticed however that it brought along some submodule stuff that I believe should not result in "changes" to develop. So I'll need to look into how to easily remove those edits from this particular PR. Any idea?

antirotor commented:
> > I've created [BigRoy#6](https://github.com/BigRoy/OpenPype/pull/6) with small changes that allows to use it on "current" system.
> 
> @antirotor Thanks - I've merged it. I have noticed however that it brought along some submodule stuff that I believe should not result in "changes" to develop. So I'll need to look into how to easily remove those edits from this particular PR. Any idea?

oh these can be just removed by another commit I guess

m-u-r-p-h-y reviewed (approved):
Successfully submitted lighting scene with Maya Hardware 2.0 renderer from Maya2022 to Deadline

![image](https://user-images.githubusercontent.com/352795/167843428-e2c28f1d-53eb-48fa-bd45-ed1f7ed80c14.png)

mkolar commented:
@m-u-r-p-h-y testing is showing all working fine. @BigRoy are you happy with this so we could merge it? I converted it from draft

BigRoy commented:
> oh these can be just removed by another commit I guess, sorry about that. They were dragged in from old branches and GitKraken doesn't handle submodules very gracefully in this regard.

I removed these submodules now.

> are you happy with this so we could merge it? I converted it from draft

I think now it should be fine. 👍

BigRoy commented on openpype/hosts/maya/plugins/publish/validate_rendersettings.py:
```diff
@@ -50,15 +50,17 @@ class ValidateRenderSettings(pyblish.api.InstancePlugin):
     ImagePrefixTokens = {
-
-        'arnold': 'maya/<Scene>/<RenderLayer>/<RenderLayer>{aov_separator}<RenderPass>',  # noqa
+        'mentalray': 'maya/<Scene>/<RenderLayer>/<RenderLayer>{aov_separator}<RenderPass>',  # noqa: E501
```
It's a bit odd however that this is a NEW addition in this PR?

antirotor replied:
this was just added so it matches the one in create render. Should be pulled from one place imho, but that can be refactored later if needed.
